REST-API-Basics/service/src/main/java/com/epam/esm/service/service/impl/GiftCertificateServiceImpl.java

```diff
@@ -0,0 +1,212 @@
+        Optional<GiftCertificate> giftCertificateOptional = giftCertificateDao.findById(id);
+        if (!giftCertificateOptional.isPresent()) {
+            throw new ServiceException("message.giftCertificateDoesntExist");
+        }
+        giftCertificateDao.deleteById(id);
+    }
+
+    @Override
+    public GiftCertificate findById(Long id) throws ServiceException {
+        Optional<GiftCertificate> giftCertificateOptional = giftCertificateDao.findById(id);
+        if (!giftCertificateOptional.isPresent()) {
+            throw new ServiceException("message.notFound");
+        }
+        return giftCertificateOptional.get();
+    }
+
+    @Override
+    public List<GiftCertificate> getAllByTagName(String tagName) {
+        return giftCertificateDao.getGiftCertificateByTagName(tagName);
+    }
+
+    @Override
+    public List<GiftCertificate> getAllWithSorting(SortParamsContext sortParamsContext) {
+        return giftCertificateDao.getAllWithSorting(sortParamsContext);
+    }
+
+    @Override
+    public List<GiftCertificate> findWithFiltering(String name, String description) {
+        return giftCertificateDao.findWithFiltering(name, description);
+    }
+
+    @Transactional
+    @Override
+    public GiftCertificateDto updateById(Long giftCertificateId, GiftCertificateDto giftCertificateDto) throws ServiceException {
+        GiftCertificate giftCertificate = giftCertificateDto.getGiftCertificate();
+        if (giftCertificate != null) {
+            if (!giftCertificateDao.findById(giftCertificateId).isPresent()) {
+                throw new ServiceException("message.cantUpdateGiftCertificateById");
+            }
+            giftCertificateDao.updateById(giftCertificateId, updateInfo(giftCertificate));
+        }
+        List<Tag> tagList = giftCertificateDto.getCertificateTags();
+        if (tagList != null) {
+            updateCertificateTags(tagList, giftCertificateId);
+        }
+        return buildGiftCertificateDto(giftCertificateDao.findById(giftCertificateId).get());
+    }
+
+    private Map<String, Object> updateInfo(GiftCertificate giftCertificate) throws ServiceException {
+        Map<String, Object> updateInfo = new HashMap<>();
+        GiftCertificateValidatorImpl giftCertificateValidator =
+                (GiftCertificateValidatorImpl) this.giftCertificateValidator;
+        String name = giftCertificate.getName();
+        if (name != null) {
+            if (!giftCertificateValidator.isNameValid(name)) {
+                throw new ServiceException("message.cantUpdateInvalidName");
+            }
+            updateInfo.put("name", name);
+        }
+        String description = giftCertificate.getDescription();
+        if (description != null) {
+            if (!giftCertificateValidator.isDescriptionValid(description)) {
+                throw new ServiceException("message.cantUpdateInvalidDescription");
+            }
+            updateInfo.put("description", description);
+        }
```

Comment:
Method can be simplified by moving parts like this to separate methods. Use IDEA hotkeys to do it quick and simple.